### PR TITLE
Lua_api.txt: Improve Lua entity rotation documentation

### LINKED ISSUE #8868: object:set_rotation: Pitch and roll are flipped?
##### Minetest version
```
5.0.1
```

##### Summary
It seems that the documentation about the object `set_rotation` function is wrong. I'm not 100% sure about this.

The pitch and roll seem to be flipped. What is called “pitch” in docs is actually roll and vice-versa.

> * `set_rotation(rot)`
>     * `rot` is a vector (radians). X is pitch (elevation), Y is yaw (heading)
>       and Z is roll (bank).

##### Steps to reproduce
* Call `set_rotation` on any object. Change the “pitch”, and then change the “roll” and see in which direction the object actually rotates.

### PR DESCRIPTION
**Pull request by paramat**
Closes #8868 
@DevyHeavy is this acceptable?

### CODE REVIEW

File: doc/lua_api.txt [hunk 1]
```diff
@@ -5470,8 +5470,14 @@ This is basically a reference to a C++ `ServerActiveObject`
     * `acc` is a vector
 * `get_acceleration()`: returns the acceleration, a vector
 * `set_rotation(rot)`
-    * `rot` is a vector (radians). X is pitch (elevation), Y is yaw (heading)
-      and Z is roll (bank).
+    * `rot` is a vector (radians).
+    * `rot.y` is rotation around the world Y axis, and acts like `set_yaw()`.
+    * `rot.x` is rotation around the object's left-right axis, so can be considered
```
Comment (ClobberXD) [inline doc/lua_api.txt]: It'd be nicer to list the components in alphabetical order. i.e. x, y, z

**Comment by DS-Minetest** [comment]
If one knows about euler angles, such a detailed description isn't needed.

File: doc/lua_api.txt [hunk 2]
```diff
@@ -5470,8 +5470,14 @@ This is basically a reference to a C++ `ServerActiveObject`
...
+      to be 'pitch'. The object's left-right axis is rotated around the world Y axis
+      by rot.y (yaw).
+    * `rot.z is rotation around the object's front-back axis, so can be considered
+      to be 'roll'. The object's front-back axis is rotated around the object's
+      left-right axis by rot.x (pitch).
```
Comment (DS-Minetest) [inline doc/lua_api.txt]: The object is first rotated around the z-axis by `rot.z` (roll), then around the x-axis by `rot.x` (pitch) and lastly around the y-axis by `rot.y` (yaw). All these rotations are counter-clock-wise, I think.
The "front-back" axis is also rotated by yaw, not only by pitch.
The front can be considered to be in (-?)z direction.

File: doc/lua_api.txt (hunk 1 above)
Comment (paramat) [reply doc/lua_api.txt]: The order is intentional and explained in the issue https://github.com/minetest/minetest/issues/8868#issuecomment-526355920
How rotation works is far clearer when understood in the order YXZ, because that's the order of the application of the rotation components, and the order of what axes are rotated by what other axes.
I won't change this order as it's logical and essential for clarity

Y coming first is also consistent with how previously we only had 'set yaw', and how the effect of yaw is unchanged and has priority.

File: doc/lua_api.txt (hunk 2 above)
Comment (paramat) [reply doc/lua_api.txt]: Well, you have the correct sequence but backwards, maybe because you are understanding it in a different way.
It's clearer to understand it 'from world to object' in terms of a gimbal system.
> The "front-back" axis is also rotated by yaw, not only by pitch.

Yes, but that's implicit in
`The object's front-back axis is rotated around the object's left-right axis`
because
`The object's left-right axis is rotated around the world Y axis by rot.y'

**Comment by paramat** [comment]
Most people don't understand Euler angles, 'extrinsic' or 'intrinsic'.

**Comment by DS-Minetest** [comment]
> Most people don't understand Euler angles, 'extrinsic' or 'intrinsic'.

Then link to wikipedia. The minetest lua reference is not responsible for explaining this.

File: doc/lua_api.txt (hunk 2 above)
Comment (DS-Minetest) [reply doc/lua_api.txt]: Well I could write: _Your intrinsic sequence is backwards!_ \:P

> > The "front-back" axis is also rotated by yaw, not only by pitch.
> 
> Yes, but that's implicit in
> `The object's front-back axis is rotated around the object's left-right axis`
> because
> `The object's left-right axis is rotated around the world Y axis by rot.y`

No it isn't implicit. Let's take an example:
yaw: 90°, pitch: 45°
(initial front: (0,0,-1), initial left: (1,0,0), ccw rotation)
after yaw: front: (1,0,0), left: (0,0,1)
after pitch (rotation around (0,0,1)): front: (sqrt(2)/2,sqrt(2)/2,0) (= the actual roll axis), rotated front-back axis (= rotated initial front): still (0,0,-1)
TL;DR: Rotation around a rotated axis is not the same as two rotations.

Comment (paramat) [reply doc/lua_api.txt]: That doesn't make sense to me.

**Comment by paramat** [comment]
Yes, the documentation is responsible for giving users a reasonable basic understanding, this is why the issue was opened in the first place because the current documentation is not clear enough.
Mentioning 'euler', 'extrinsic' or 'intrinsic' is not essential, but we could do that too, i don't mind. I was objecting to doing that 'instead of ...' as you suggested.

File: doc/lua_api.txt (hunk 1 above)
Comment (ClobberXD) [reply doc/lua_api.txt]: Order of application of rotation deserves to explicitly mentioned, IMO.

Comment (paramat) [reply doc/lua_api.txt]: It is, although maybe not as clearly as it could be:
```
The object's left-right axis is rotated around the world Y axis by rot.y
The object's front-back axis is rotated around the object's left-right axis
```

Comment (ClobberXD) [reply doc/lua_api.txt]: No, I meant that the order of application of rotation should be mentioned as a separate point, and the components should be documented in alphabetical order. Modders are used to X first, then Y and Z, and it'd be more intuitive for them if the documentation lists them in correct order. And as you said, the ordering of the lines doesn't offer any clue as to what the order of rotation is.

Comment (paramat) [reply doc/lua_api.txt]: I need to improve this (see below).
The explanation should hopefully explain what you mean by 'order', it may be clearer if it is not a separate point but integrated into the explanation, as the current form of the PR attempts to.
> And as you said, the ordering of the lines doesn't offer any clue as to what the order of rotation is.

I didn't write that, i actually wrote the opposite, see "it is" above, i just wrote that it could be clearer.
I also made it clear that the YXZ component order of the documentation is specifically intended to explain the 'order of rotation'.
I'm bemused why you would claim i wrote something that is so clearly opposite to what i actually wrote.

I've also explained why the YXZ component order of the documentation is essential, and given several good reasons. Ordering for the sake of being alphabetical is a non-argument. We're used to XYZ elsewhere because it makes sense elsewhere, but here it doesn't. Here, YXZ will be more intuitive for those trying to understand these rotations.

File: doc/lua_api.txt (hunk 2 above)
Comment (paramat) [reply doc/lua_api.txt]: Woahh ... i just realised a mistake, the object left-right axis rotates when roll is applied, so i can't use that, instead i have to write about the 'pitch axis' which is different.
Your comments helped me think further.
So WIP, will try to improve.

**Comment by paramat** [comment]
WIP because there is a mistake in the explanation, and it can be clearer, i'll improve it.

I encourage any reviewers use this modified boat code, if they haven't already, to help understand these rotations:
```
function boat.on_step(self, dtime)
	if self.driver then
		local driver_objref = minetest.get_player_by_name(self.driver)
		if driver_objref then
			local prot = self.object:get_rotation()
			print("X rot " .. prot.x)
			print("Y rot " .. prot.y)
			print("Z rot " .. prot.z)
			print("")
			local ctrl = driver_objref:get_player_control()
			local rot = self.object:get_rotation()
			-- Yaw
			if ctrl.left then
				rot.y = rot.y + dtime * 1
			elseif ctrl.right then
				rot.y = rot.y - dtime * 1
			end
			-- Pitch
			if ctrl.up then
				rot.x = rot.x + dtime * 1
			elseif ctrl.down then
				rot.x = rot.x - dtime * 1
			end
			-- Roll
			if ctrl.jump then
				rot.z = rot.z + dtime * 1
			elseif ctrl.sneak then
				rot.z = rot.z - dtime * 1
			end
			self.object:set_rotation(rot)
			self.object:set_pos(self.object:get_pos())
		end
	end
end
```
First try yaw and pitch only, to get a feel for those.
Then at a particular yaw and pitch, try roll, but return roll to 0 before continuing, otherwise it gets really confusing.

You'll discover:
Yaw (rot.y) rotates the entire rotational system around the world Y-axis so is primary, it rotates the pitch (rot.x) axis which lies within a horizontal plane and is secondary.
The object front-back axis, the roll (rot.z) axis, is tertiary and rotates around the pitching axis.

File: doc/lua_api.txt (hunk 1 above)
Comment (ClobberXD) [reply doc/lua_api.txt]: > I didn't write that, i actually wrote the opposite, see "it is" above, i just wrote that it could be clearer.

That it could be made clearer is what I was referring to, sorry for the confusion. :)

> I also made it clear that the YXZ component order of the documentation is specifically intended to explain the 'order of rotation'.

The order of documentation is **implicit**, and isn't very helpful in indicating the order of rotation. As a modder, the first thing that'd come to my mind when I look at this documentation is that the order is messed up, and I would not have any way of knowing that the order of documentation of the components corresponds to the order of application of rotation. This is because there's currently nothing in the documentation that indicates that the points are documented in a "sequence".

Inclusion of words/phrases like "First, the object is rotated in the ... axis", and "Then the object is rotated in the ... axis" would indicate the existence of a sequence. But then the documentation would become unnecessarily verbose, which is why I suggested **explicitly** mentioning the order of application of rotations. e.g.

```
  * Note: Rotations are applied in the following order: Y -> X -> Z
  * <Document X> 
  * <Document Z> 
  * <Document Y>
```

**Comment by DS-Minetest** [comment]
>Then at a particular yaw and pitch, try roll, but return roll to 0 before continuing, otherwise it gets really confusing.

Is it possible that you don't use the +z-axis as forward direction? In my test code posted at https://github.com/minetest/minetest/issues/8868#issuecomment-526399859 I'm continuously increasing roll on step and it's not confusing.

Btw. all the rotations are right-handed rotations around the +(x/y/z)axes. (If you look in -(x/y/z) direction, it's ccw.)
I think, this should be documented.

**Comment by paramat** [comment]
The boat rotation seems correct, i'm just warning about how it can get complex when roll gets involved.
I have no idea if the boat front is +z or -z.

I'm rather unsure now about altering the documentation, as the current documentation is correct, and this PR is difficult.

**Comment by Wuzzy2** [comment]
I like the new text more. Thanks.
